### PR TITLE
Move one level up in layer hierarchy

### PR DESCRIPTION
**Pull request by teunbrand**
Hi there,

This PR aims to improve compatibility with {ggblend}.
Briefly, it changes the location where the rasterisation information is appended to grobs from the `Geom$draw_panel()` level to the `Geom$draw_layer()` level. This should have no drawbacks for {ggrastr}, but makes it compatible with {ggblend}. See also https://github.com/mjskay/ggblend/issues/18.

Best,
Teun

### CODE REVIEW

**Review by evanbiederstedt — APPROVED** [review]
LGTM

@teunbrand My sole question here is whether this change deserves a new version on CRAN. 

If so, could you create an updated version of the package and update the details in the CHANGELOG within this PR? That would be helpful

**Comment by mjskay** [comment]
Hmm I'm not sure if this is working correctly? I would have thought it would need to move to the level of `Layer$draw_geom()` to work with ggblend, so I was surprised when it worked just at the level of `Geom$draw_layer` --- and looking at the output, I think it isn't being rasterized at all now?

**Comment by evanbiederstedt** [comment]
Do you both have some example commands to check? @teunbrand @mjskay 

Working with a few examples would be useful. 

@teunbrand 
> This should have no drawbacks for {ggrastr}, but makes it compatible with {ggblend}. 

Could you detail this a bit more? 

I should clarify my PR review was based on just looking at the code; nothing in set in stone so far.

**Comment by teunbrand** [comment]
Thanks for catching this @mjskay! Indeed, it seems we need this from the geom level up to the layer level. Another issue that came up during testing was that essentially none of the devices supported by {ggrastr} allow the new R4.1+ graphics features, so I had to add a `dev = "cairo_png"` option. However, it works now:

``` r
library(ggplot2)
library(ggblend)
devtools::load_all("~/packages/ggrastr/")
#> ℹ Loading ggrastr

file <- "temp.png"
png(file)

diamonds |>
  ggplot() + 
  geom_point(aes(carat, price, colour = cut)) |>  blend("multiply") |> 
  rasterize(dpi = 30, dev = "cairo_png")
#> Warning: Your graphics device, "png", reports that blend = "multiply" is not supported.
#>  - If the blending output IS NOT as expected (e.g. geoms are not being
#>    drawn), then you must switch to a graphics device that supports
#>    blending, like png(type = "cairo"), svg(), or cairo_pdf().
#>  - If the blending output IS as expected despite this warning, this is
#>    likely a bug *in the graphics device*. Unfortunately, several
#>    graphics do not correctly report their capabilities. You may wish to
#>    a report a bug to the authors of the graphics device. In the mean
#>    time, you can disable this warning via options(ggblend.check_blend =
#>    FALSE).
#>  - For more information, see the Supported Devices section of
#>    help('blend').

dev.off()
#> png 
#>   2
knitr::include_graphics(file)
```



<sup>Created on 2023-05-22 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

The warning that {ggblend} emits is hard to avoid here, I think. This is probably due to looking at the global device, rather than the specific device that is used to rasterise the layer.

> Could you detail this a bit more?

There were no user-visible changes that his PR would introduce is what I meant. However, now there is a new device option.

> I should clarify my PR review was based on just looking at the code; nothing in set in stone so far.

Of course, I wouldn't want to merge bad code if we could catch it early.

**Comment by mjskay** [comment]
Awesome, thanks so much!

Looks like cairo_png is another graphics device that isn't reporting blend capabilities correctly, hah ... :/

**Comment by teunbrand** [comment]
> Looks like cairo_png is another graphics device that isn't reporting blend capabilities correctly

I think it does report it correctly, it is just that {ggrastr} makes a temporary device for the purposes of rasterisation and then switches back to whatever device was active before, which happens at the `makeContent`/drawing step. The {ggblend} warning is about the active device, not the temporary one.

**Comment by mjskay** [comment]
Ah, makes sense. Perhaps I should move the warning into a grob that wraps the layer's grobs or something.

File: R/rasterise.R [hunk 1]
```diff
@@ -173,12 +180,12 @@ makeContext.rasteriser <- function(x) {
   grid::popViewport()
 
   # Capture raster
-  if (dev != "ragg_png") {
+  if (!dev %in% c("ragg_png", "cairo_png")) {
```
Comment (evanbiederstedt) [inline R/rasterise.R:186]: Thanks for doing this anyways, this is the correct syntax

**Review by evanbiederstedt — APPROVED** [review]
LGTM

**Comment by evanbiederstedt** [comment]
Doublechecking: 

@mjskay @teunbrand 
You're happy with the PR and all conversations resolved? Ready for a merge?

**Comment by teunbrand** [comment]
Yeah I think this PR is in a good shape. However, there were some R CMD check flags that are unrelated to this PR that might be worth investigating at some point.

**Comment by mjskay** [comment]
Yup looks good to me, thanks!

**Comment by evanbiederstedt** [comment]
Should be fixed for CRAN. Please have another look of course. 

I'll update the version to 1.0.2

**Comment by evanbiederstedt** [comment]
New version now on CRAN
